pkg/codegen/schema/schema.go

```diff
@@ -0,0 +1,898 @@
+	Properties []*Property
+	// StateInputs is the set of inputs used to get an existing resource, if any.
+	StateInputs *ObjectType
+	// Aliases is the list of aliases for the resource.
+	Aliases []*Alias
+	// DeprecationMessage indicates whether or not the resource is deprecated.
+	DeprecationMessage string
+	// Language specifies additional language-specific data about the resource.
+	Language map[string]json.RawMessage
+}
+
+// Function describes a Pulumi function.
+type Function struct {
+	// Token is the function's Pulumi type token.
+	Token string
+	// Comment is the description of the function, if any.
+	Comment string
+	// Inputs is the bag of input values for the function, if any.
+	Inputs *ObjectType
+	// Outputs is the bag of output values for the function, if any.
+	Outputs *ObjectType
+	// DeprecationMessage indicates whether or not the function is deprecated.
+	DeprecationMessage string
+	// Language specifies additional language-specific data about the function.
+	Language map[string]json.RawMessage
+}
+
+// Package describes a Pulumi package.
+type Package struct {
+	moduleFormat *regexp.Regexp
+
+	// Name is the unqualified name of the package (e.g. "aws", "azure", "gcp", "kubernetes". "random")
+	Name string
+	// Version is the version of the package.
+	Version *semver.Version
+	// Description is the description of the package.
+	Description string
+	// Keywords is the list of keywords that are associated with the package, if any.
+	Keywords []string
+	// Homepage is the package's homepage.
+	Homepage string
+	// License indicates which license is used for the package's contents.
+	License string
+	// Repository is the URL at which the source for the package can be found.
+	Repository string
+	// LogoURL is the URL for the package's logo, if any.
+	LogoURL string
+
+	// Types is the list of non-resource types defined by the package.
+	Types []Type
+	// Config is the set of configuration properties defined by the package.
+	Config []*Property
+	// Provider is the resource provider for the package, if any.
+	Provider *Resource
+	// Resources is the list of resource types defined by the package.
+	Resources []*Resource
+	// Functions is the list of functions defined by the package.
+	Functions []*Function
+	// Language specifies additional language-specific data about the package.
+	Language map[string]json.RawMessage
+}
+
+func (pkg *Package) TokenToModule(tok string) string {
+	// token := pkg ":" module ":" member
+
+	components := strings.Split(tok, ":")
+	if len(components) != 3 {
+		return ""
+	}
+
+	matches := pkg.moduleFormat.FindStringSubmatch(components[1])
+	if len(matches) < 2 || matches[1] == "index" {
+		return ""
+	}
+	return matches[1]
+}
+
+// TypeSpec is the serializable form of a reference to a type.
+type TypeSpec struct {
+	// Type is the primitive or composite type, if any. May be "bool", "integer", "number", "string", "array", or
+	// "object".
+	Type string `json:"type,omitempty"`
```

Comment:
omitempty makes it sound like this is optional.  is it?  or is this required?  if required, is there a way to ndicate htat?

Reply:
It is not required: refs and oneof omit this field.